### PR TITLE
[az fleet] Added and updated identity related commands & parameters

### PR DESCRIPTION
**Pull request by Ealianis**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
Changes to : `az fleet create`, `az fleet update`, and new `az fleet identity `.

* Added fleet identity sub command group.
* Updated identity parameter names to match design guidelines.
* Updated help text, and tests.

Adding support for 

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
az fleet

File: src/fleet/azext_fleet/custom.py [hunk 1]
```diff
@@ -212,6 +216,89 @@ def get_credentials(cmd,  # pylint: disable=unused-argument
...
+        operation_group="fleets"
+    )
+    fleet_managed_service_identity_model = cmd.get_models(
+        "ManagedServiceIdentity",
+        resource_type=CUSTOM_MGMT_FLEET,
+        operation_group="fleets"
+    )
+
+    if system_assigned is False and user_assigned is False:
+        raise CLIError("Either --system-assigned is set to true, or pass an empty --user-assigned")
```
Comment (jim-minter) [inline src/fleet/azext_fleet/custom.py:281]: please check the error message?

Comment (jim-minter) [inline src/fleet/azext_fleet/custom.py:281]: Unfortunately, I think you need to go an extra mile here:

if --system-assigned is set, you should get the fleet, validate it's system assigned, then set it to None
if --user-assigned is set, you should get the fleet, validate it's user assigned with the identity provided, then set it to None

e.g.:
- remove --user-assigned should **fail** to set a system-assigned to None
- remove --user-assigned foo should **fail** to set a user-assigned bar to None

File: src/fleet/azext_fleet/custom.py [hunk 2]
```diff
@@ -212,6 +216,89 @@ def get_credentials(cmd,  # pylint: disable=unused-argument
...
+
+    return sdk_no_wait(no_wait, client.begin_update, resource_group_name, name, fleet_patch, polling_interval=5)
+
+
+def show_fleet_identity(cmd,  # pylint: disable=unused-argument
+               client,
+               resource_group_name,
+               name):
+    
+    identity = getattr(client.get(resource_group_name, name), 'identity', [])
```
Comment (jim-minter) [inline src/fleet/azext_fleet/custom.py:297]: should be `{}`? or is `[]` correct?

**Review by jim-minter — CHANGES_REQUESTED** [review]
Smallish suggestions, then will also need tests please.

File: src/fleet/azext_fleet/custom.py [hunk 3]
```diff
@@ -95,19 +95,21 @@ def create_fleet(cmd,
         resource_type=CUSTOM_MGMT_FLEET,
         operation_group="fleets"
     )
+    
+    # Handle all 6 permutations of mi_system & mi_user assigned parameters
```
[the remark this answers is not in the document]
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py]: Simplified

File: src/fleet/azext_fleet/custom.py (hunk 1 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py]: Updated.

File: src/fleet/azext_fleet/custom.py [hunk 4]
```diff
@@ -143,22 +145,24 @@ def update_fleet(cmd,
         operation_group="fleets"
     )
 
-    if enable_managed_identity is None:
+    # Handle all 6 permutations of mi_system & mi_user assigned parameters
```
[the remark this answers is not in the document]
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py]: Simplified

File: src/fleet/azext_fleet/custom.py (hunk 2 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py]: Yup {}, not an array. The guidelines brought to attention that resources can have multiple identities, which we obviously don't support ATM.

File: src/fleet/azext_fleet/custom.py (hunk 1 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py]: Updated to get and check.

File: src/fleet/azext_fleet/_params.py [hunk 5]
```diff
@@ -29,20 +29,31 @@ def load_arguments(self, _):
...
         c.argument('apiserver_subnet_id', validator=validate_apiserver_subnet_id, is_preview=True, help='The subnet to be used when apiserver vnet integration is enabled. It is required when creating a new Fleet with BYO vnet.')
         c.argument('agent_subnet_id', validator=validate_agent_subnet_id, is_preview=True, help='The ID of the subnet which the Fleet hub node will join on startup. If this is not specified, a vnet and subnet will be generated and used.')
-        c.argument('enable_managed_identity', action='store_true', help='Enable system assigned managed identity (MSI) on the Fleet resource.')
-        c.argument('assign_identity', validator=validate_assign_identity, help='With --enable-managed-identity, enable user assigned managed identity (MSI) on the Fleet resource. Specify the existing user assigned identity resource.')
+        c.argument('mi_system_assigned', action='store_true', help='Enable system assigned managed identity (MSI) on the Fleet resource.')
+        c.argument('mi_user_assigned', validator=validate_identity_id, help='The resource ID of an existing user assigned identity.')
```
Comment (jim-minter) [inline src/fleet/azext_fleet/_params.py:33]: Does this help text make sense?

File: src/fleet/azext_fleet/_params.py [hunk 6]
```diff
@@ -29,20 +29,31 @@ def load_arguments(self, _):
...
-        c.argument('assign_identity', validator=validate_assign_identity, help='With --enable-managed-identity, enable user assigned managed identity (MSI) on the Fleet resource. Specify the existing user assigned identity resource.')
+        c.argument('mi_system_assigned', arg_type=get_three_state_flag(), help='Enable system assigned managed identity (MSI) on the Fleet resource.')
+        c.argument('mi_user_assigned', validator=validate_identity_id, help='The resource ID of an existing user assigned identity.')
 
     with self.argument_context('fleet get-credentials') as c:
         c.argument('context_name', options_list=['--context'], help='If specified, overwrite the default context name.')
         c.argument('path', options_list=['--file', '-f'], type=file_type, completer=FilesCompleter(), default=os.path.join(os.path.expanduser('~'), '.kube', 'config'))
 
+    with self.argument_context('fleet identity') as c:
+        c.argument('name', options_list=['--fleet-name', '-f'], help='Specify the fleet name.')
```
Comment (jim-minter) [inline src/fleet/azext_fleet/_params.py:47]: I think that when dealing with a top-level resource, it's always `-n`?  `-f` for the fleet name when dealing with a subresource?

File: src/fleet/azext_fleet/_help.py [hunk 7]
```diff
@@ -313,3 +313,58 @@
     short-summary: Wait for a update strategy resource to reach a desired state.
     long-summary: If an operation on an update strategy was interrupted or was started with `--no-wait`, use this command to wait for it to complete.
 """
+
+helps['fleet identity'] = """
+    type: group
+    short-summary: Commands to the fleet identity.
```
Comment (jim-minter) [inline src/fleet/azext_fleet/_help.py:319]: Check grammar

File: src/fleet/azext_fleet/_params.py (hunk 5 above)
Comment (jim-minter) [reply src/fleet/azext_fleet/_params.py:33]: E.g. don't we want something like 'Enable user managed identity ... with this resource ID'

File: src/fleet/azext_fleet/custom.py [hunk 8]
```diff
@@ -143,22 +144,21 @@ def update_fleet(cmd,
...
-            managed_service_identity.user_assigned_identities = {assign_identity: user_assigned_identity_model()}
+    elif mi_user_assigned is not None:
+        managed_service_identity = fleet_managed_service_identity_model(type="UserAssigned")
+        user_assigned_identity_model = cmd.get_models(
+            "UserAssignedIdentity",
+            resource_type=CUSTOM_MGMT_FLEET,
+            operation_group="fleets")
+        managed_service_identity.user_assigned_identities = {mi_user_assigned: user_assigned_identity_model()}
+    else:
+        managed_service_identity = fleet_managed_service_identity_model(type="None")
```
Comment (jim-minter) [inline src/fleet/azext_fleet/custom.py:161]: Won't this remove by default, despite the comment above?

Why does the update code use if / elif and the create code use if / if ?  Be consistent.

File: src/fleet/azext_fleet/custom.py [hunk 9]
```diff
@@ -212,6 +212,97 @@ def get_credentials(cmd,  # pylint: disable=unused-argument
         raise CLIError("Fail to find kubeconfig file.")
 
 
+def assign_fleet_identity(cmd,
```
Comment (jim-minter) [inline src/fleet/azext_fleet/custom.py:215]: I think there is significant trouble here.  Let's discuss on a call.

**Review by jim-minter — CHANGES_REQUESTED** [review]

File: src/fleet/azext_fleet/_params.py (hunk 6 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/_params.py:47]: Good point, that makes more sense than what I had in mind.
Reverted.

File: src/fleet/azext_fleet/_params.py (hunk 5 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/_params.py:33]: Improved.

File: src/fleet/azext_fleet/_help.py (hunk 7 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/_help.py:319]: Fixed.

File: src/fleet/azext_fleet/custom.py (hunk 8 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py:161]: reverted back to elseif.

**Comment by Ealianis** [comment]
We are closing this PR as the design guidelines do not quite have the adoption or consideration that works for our APIs ATM.

File: src/fleet/azext_fleet/custom.py (hunk 9 above)
Comment (Ealianis) [reply src/fleet/azext_fleet/custom.py:216]: Discussed, closing PR as discussed. Tabling az identity sub command changes.

**Pull request closed without merging**